### ISSUE TITLE
Model saving timeout

### ISSUE THREAD

**Issue by pierrenodet** [MGabr/glint-word2vec#1]
Hi, thanks for this library !

I have an issue when trying to save a big model (vocab size of 10 millions), I'm hitting the 5 minutes timeout of this line of code :

```scala    
val savedFuture = matrix.save(path, sc.hadoopConfiguration)
Await.ready(savedFuture, 5 minutes)
```

Is it an issue on my side or should the 5 minutes be configurable ?

**Comment by MGabr** [comment]
Hi,

a vocab size of 10 million should probably not be the problem. The evaluation I did (https://github.com/MGabr/evaluate-glint-word2vec/blob/master/project_report.pdf) was with a vocab size of 8.7 million.

Which number of parameter servers and word vector dimensions are you using?
Did you start the parameter servers as separate Spark application?

Best regards,
Markus

**Comment by pierrenodet** [comment]
I tried to follow the setup from your report, which is a parameter server launch in a separate Spark application with 5 executors of 16 cores and 20g of RAM. Then I launched another app with 20 executors of 16 cores and 20g of RAM.

I used 5 parameter servers and a vector size of 100.

Just to be sure in order to use the separate parameter server i just need to pass the IP address of the driver from the spark app containing the parameter server to the model as a string ? Or do you I need to add a port ?
```scala
val word2vec = new ServerSideGlintWord2Vec()
.setParameterServerHost("GLINT.DRIVER.IP.ADDRESS")
.setNumParameters(5)
```

**Comment by MGabr** [comment]
Ok, your settings seem to be fine. The IP address is also set correctly and if there was an issue with it there would be an error earlier on before save.

I had an issue in another project using the parameter servers (not Word2Vec) where save timed out but other requests to the parameter servers still worked. I haven't been able to fix this. I'm not sure if this is related though.

In any case, parameter server errors are not passed on, so the timeout may also be caused by the servers failing to save to the given HDFS path. Does saving a model with a small vocab size to the same path work for you?

**Comment by pierrenodet** [comment]
Yeah when I setup a higher mincount to reduce my vocabulary (2 million), my model was perfectly saved to hdfs (the matrix and the vocab). With my complete vocab, only the vocab is stored in the hdfs when saving the model.

Thanks for helping me ! I'm gonna try again this weekend.

**Comment by MGabr** [comment]
Yeah, then I fear this is the same problem as in my other project. If a vocab size of 10 million and a dimension of 100 is all you need you can still call `model.toLocal` to pull all the word vectors from the parameter servers and create a common Spark Word2Vec model with them. This would even be a cleaner approach since you then don't need parameter servers at transformation time, but this only works for word vector matrices of size up to 8GB due to Spark's broadcast limit and you need enough memory on the executors.

For larger matrices I might this weekend add an experimental model which pulls the word vectors into an RDD and joins this RDD at transformation time. I wouldn't be able to check the performance of this join approach at the moment though.

**Comment by pierrenodet** [comment]
I'm gonna try that. Thanks for the help ! Nice library ! 

Indeed having a RDD instead of a broadcasted model will be slower but at least it's gonna scale to bigger model.
